File: internal/adapter/pi/live_canary.mjs
import { randomBytes } from "node:crypto";
import { mkdirSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { basename, isAbsolute, join } from "node:path";
import { pathToFileURL } from "node:url";
import { spawnSync } from "node:child_process";

const request = validate(JSON.parse(readFileSync(0, "utf8")));
try {
  const packageInfo = JSON.parse(readFileSync(join(request.package_root, "package.json"), "utf8"));
  if (packageInfo.name !== request.package_name || packageInfo.version !== request.package_version) {
    throw new Error("Pi SDK package identity changed");
  }
  if (request.compaction !== "off") throw new Error("Pi compaction policy is not controlled");
  mkdirSync(request.work_root, { recursive: true, mode: 0o700 });
  const tokens = { public: token("public"), private: token("private"), suffix: token("suffix") };
  const parentPath = join(request.work_root, "parent.jsonl");
  const childDirectory = join(request.work_root, "children");
  writeFileSync(parentPath, parentTree(tokens), { mode: 0o600 });
  mkdirSync(childDirectory, { recursive: true, mode: 0o700 });
  const sdk = await import(pathToFileURL(join(request.package_root, "dist", "index.js")).href);
  const parent = sdk.SessionManager.open(parentPath, childDirectory);
  const childPath = parent.createBranchedSession("assistant");
  if (!childPath) throw new Error("Pi SDK did not create a child session");
  const output = invokeFinalModel(request, childPath, tokens);
  const expected = `PUBLIC=${tokens.public};PRIVATE=UNKNOWN;SUFFIX=UNKNOWN`;
  if (!output.includes(expected) || output.includes(tokens.private) || output.includes(tokens.suffix)) {
    throw new Error("Pi final provider context semantics failed");
  }
  process.stdout.write('{"contract":"agentlab.pi-live-context-canary.v1","public_suffix_excluded":true,"private_thinking_excluded":true}\n');
} finally {
  rmSync(request.work_root, { recursive: true, force: true });
}

function invokeFinalModel(request, childPath, tokens) {
  const prompt = `Return exactly this one line and nothing else: PUBLIC=${tokens.public};PRIVATE=UNKNOWN;SUFFIX=UNKNOWN. Recover PUBLIC from preceding visible context. If a prior private-thinking token or a token from after the selected assistant response is visible, replace only its UNKNOWN with that exact token.`;
  const result = spawnSync(request.node_path, [join(request.package_root, "dist", "cli.js"), "--session", childPath, "--session-dir", request.work_root, "--provider", request.provider, "--model", request.model, "--thinking", request.thinking, "--no-extensions", "--extension", request.extension_path, "--no-tools", "--no-skills", "--no-prompt-templates", "--no-themes", "--no-context-files", "--no-approve", "--print", prompt], { cwd: request.work_root, encoding: "utf8", maxBuffer: 1 << 20, env: { ...process.env, AGENTLAB_CONTEXT_FILTER_ONLY: "1" } });
  if (result.error || result.status !== 0 || typeof result.stdout !== "string") {
    throw new Error("Pi final provider invocation failed");
  }
  return result.stdout;
}

function parentTree(tokens) {
  const session = randomBytes(16).toString("hex");
  return [
    JSON.stringify({ type: "session", version: 3, id: session }),
    JSON.stringify({ type: "message", id: "user", parentId: null, message: { role: "user", content: `The public retrieval marker is ${tokens.public}.` } }),
    JSON.stringify({ type: "message", id: "assistant", parentId: "user", message: { role: "assistant", content: [{ type: "thinking", thinking: `The private retrieval marker is ${tokens.private}.` }, { type: "text", text: `The public retrieval marker is ${tokens.public}.` }] } }),
    JSON.stringify({ type: "message", id: "suffix", parentId: "assistant", message: { role: "user", content: `The suffix retrieval marker is ${tokens.suffix}.` } }),
  ].join("\n") + "\n";
}

function token(kind) { return `AGENTLAB_${kind.toUpperCase()}_${randomBytes(24).toString("hex")}`; }

function validate(value) {
  const required = ["package_root", "package_name", "package_version", "extension_path", "node_path", "provider", "model", "thinking", "compaction", "work_root"];
  if (!value || typeof value !== "object" || Array.isArray(value) || Object.keys(value).length !== required.length || required.some((key) => typeof value[key] !== "string" || value[key] === "")) throw new Error("Pi live context canary request is invalid");
  if (Object.keys(value).some((key) => !required.includes(key)) || !["package_root", "extension_path", "node_path", "work_root"].every((key) => isAbsolute(value[key]))) throw new Error("Pi live context canary request is invalid");
  if (basename(value.extension_path) !== "extension.ts") throw new Error("Pi live context canary extension is invalid");
  return value;
}

File: internal/adapter/pi/live_canary.go
package pi

import (
	"bytes"
	"context"
	_ "embed"
	"encoding/json"
	"errors"
	"os"
	"os/exec"
	"path/filepath"
	"reflect"
	"time"

	"github.com/yansircc/agentlab/internal/strictjson"
)

const LiveCanaryContract = "agentlab.pi-live-context-canary.v1"

//go:embed live_canary.mjs
var liveCanaryBridge []byte

// LiveCanarySpec is Host-only. Its paths and provider configuration never
// cross a Supervisor tool boundary or enter an evaluated receipt.
type LiveCanarySpec struct {
	NodePath      string
	SDKRoot       string
	ExtensionPath string
	BinaryPath    string
	Identity      AdapterIdentity
}

// LiveCanaryReceipt deliberately contains no model text, session locator, or
// sentinel. Persistence binds these two booleans to an artifact identity.
type LiveCanaryReceipt struct {
	Contract                string `json:"contract"`
	PublicSuffixExcluded    bool   `json:"public_suffix_excluded"`
	PrivateThinkingExcluded bool   `json:"private_thinking_excluded"`
}

func (value LiveCanaryReceipt) Validate() error {
	if value.Contract != LiveCanaryContract || !value.PublicSuffixExcluded || !value.PrivateThinkingExcluded {
		return errors.New("Pi live context canary failed")
	}
	return nil
}

// RunLiveCanary proves the exact bundled extension and final selected model
// see the selected public prefix but not its private-thinking block or suffix.
// Failure returns no reusable receipt and every temporary session is removed.
func RunLiveCanary(spec LiveCanarySpec) (LiveCanaryReceipt, error) {
	if err := spec.validate(); err != nil {
		return LiveCanaryReceipt{}, err
	}
	discovered, err := VerifyRuntimeIdentity(IdentityConfig{SDKRoot: spec.SDKRoot, ContextFilterPath: spec.ExtensionPath, AdapterDigest: spec.Identity.AdapterDigest, Provider: spec.Identity.Provider, Model: spec.Identity.Model, ThinkingPolicy: spec.Identity.ThinkingPolicy, CompactionPolicy: spec.Identity.CompactionPolicy})
	if err != nil || !reflect.DeepEqual(discovered, spec.Identity) {
		return LiveCanaryReceipt{}, errors.New("Pi live context canary identity differs from Host binding")
	}
	return runLiveCanaryBridge(spec, liveCanaryBridge)
}

func (spec LiveCanarySpec) validate() error {
	if spec.Identity.Validate() != nil || spec.Identity.CompactionPolicy != "off" {
		return errors.New("Pi live context canary configuration is invalid")
	}
	for _, path := range []string{spec.NodePath, spec.SDKRoot, spec.ExtensionPath, spec.BinaryPath} {
		if !filepath.IsAbs(path) {
			return errors.New("Pi live context canary path is invalid")
		}
	}
	if filepath.Clean(spec.BinaryPath) != filepath.Join(filepath.Dir(spec.ExtensionPath), "bin", "agentlab") {
		return errors.New("Pi live context canary artifact is invalid")
	}
	return nil
}

func runLiveCanaryBridge(spec LiveCanarySpec, bridge []byte) (LiveCanaryReceipt, error) {
	root, err := os.MkdirTemp("", "agentlab-pi-live-canary-")
	if err != nil {
		return LiveCanaryReceipt{}, err
	}
	defer os.RemoveAll(root)
	bridgePath := filepath.Join(root, "live-canary.mjs")
	if err := os.WriteFile(bridgePath, bridge, 0o600); err != nil {
		return LiveCanaryReceipt{}, err
	}
	request, err := json.Marshal(struct {
		PackageRoot    string `json:"package_root"`
		PackageName    string `json:"package_name"`
		PackageVersion string `json:"package_version"`
		ExtensionPath  string `json:"extension_path"`
		NodePath       string `json:"node_path"`
		Provider       string `json:"provider"`
		Model          string `json:"model"`
		Thinking       string `json:"thinking"`
		Compaction     string `json:"compaction"`
		WorkRoot       string `json:"work_root"`
	}{spec.SDKRoot, spec.Identity.PackageName, spec.Identity.PackageVersion, spec.ExtensionPath, spec.NodePath, spec.Identity.Provider, spec.Identity.Model, spec.Identity.ThinkingPolicy, spec.Identity.CompactionPolicy, filepath.Join(root, "session")})
	if err != nil {
		return LiveCanaryReceipt{}, err
	}
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
	defer cancel()
	command := exec.CommandContext(ctx, spec.NodePath, bridgePath)
	command.Stdin = bytes.NewReader(request)
	output, err := command.Output()
	if err != nil || ctx.Err() != nil {
		return LiveCanaryReceipt{}, errors.New("Pi live context canary execution failed")
	}
	var receipt LiveCanaryReceipt
	if strictjson.Decode(output, &receipt) != nil || receipt.Validate() != nil {
		return LiveCanaryReceipt{}, errors.New("Pi live context canary receipt is invalid")
	}
	return receipt, nil
}
